### DISCUSSION TITLE
HTTP Status Code에 대한 설명을 해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#205] [Network]
300번대 code들에 대해서는 자세히 말해주면 좋을 것 같아요.

**Comment by alstn113** [comment]
HTTP 상태 코드는 클라이언트가 보낸 요청에 대한 서버의 응답을 나타내는 숫자 코드입니다.
이 코드들은 5개의 범주로 나뉘며, 각 범주는 숫자로 구분됩니다. 각 범주 내에서도 다양한 세부 코드들이 존재합니다.

### 1xx (Informational)
- 요청이 수신되어 처리 중임을 나타냅니다.
- 잘 사용되지 않는다.

### 2xx (Success)
- 요청이 성공적으로 처리되었음을 나타냅니다.    
- 코드
  - 200 OK: 요청이 성공적으로 처리되었음을 나타냅니다.
  - 201 Created: 요청이 성공적으로 처리되었고, 새로운 리소스가 생성되었음을 나타냅니다.
  - 204 No Content: 요청이 성공적으로 처리되었지만, 응답 본문에 내용이 없음을 나타냅니다.

### 3xx (Redirection)
- 요청을 완료하려면 추가 동작이 필요함을 나타냅니다.
- 클라이언트가 관심 있어 하는 리소스에 대해 다른 위치를 사용하라고 말해주거나 그 리소스의 내용 대신 다른 대안 응답을 제공합니다.
- 리다이렉션: 클라이언트가 요청한 리소스가 다른 위치에 있을 때, 클라이언트를 그 위치로 보내는 것을 말합니다.
    - 영구 리다이렉션: 리소스가 영구적으로 다른 위치로 이동되었을 때 사용합니다.
    - 일시 리다이렉션: 리소스가 일시적으로 다른 위치로 이동되었을 때 사용합니다.
    - 특수 리다이렉션: 리소스가 특정 조건에 따라 다른 위치로 이동되었을 때 사용합니다.         
- 코드
  - 301 Moved Permanently: 요청한 리소스가 새로운 위치로 영구적으로 이동되었음을 나타냅니다. (영구 리다이렉션, 메서드가 GET으로 변경)
  - 302 Found: 요청한 리소스가 일시적으로 다른 위치로 이동되었음을 나타냅니다. (일시 리다이렉션, , 메서드가 GET으로 변경)
  - 304 Not Modified: 클라이언트가 마지막으로 요청한 리소스가 수정되지 않았음을 나타냅니다. (특수 리다이렉션)

### 4xx (Client Error)
- 클라이언트의 요청에 오류가 있음을 나타냅니다.         
- 코드
  - 400 Bad Request: 클라이언트의 요청이 잘못되었음을 나타냅니다.
  - 401 Unauthorized: 클라이언트가 인증되지 않았음을 나타냅니다.
  - 403 Forbidden: 클라이언트가 요청한 리소스에 접근할 권한이 없음을 나타냅니다.
  - 404 Not Found: 요청한 리소스를 찾을 수 없음을 나타냅니다.
  - 409 Conflict: 요청이 현재 리소스의 상태와 충돌이 발생했음을 나타냅니다.

### 5xx (Server Error)
- 서버가 요청을 처리하는 중에 오류가 발생했음을 나타냅니다.       
- 코드
  - 500 Internal Server Error: 서버에 오류가 발생했음을 나타냅니다.

참고: https://www.restapitutorial.com/httpstatuscodes
참고: https://inpa.tistory.com/entry/HTTP-🌐-상태-코드-1XX-5XX-총정리판-📖

**Comment by 3Juhwan** [comment]
HTTP Status Codes는 클라이언트와 서버 간의 통신에서 요청의 결과를 나타내는 3자리 숫자 코드입니다. 이 코드들은 5개의 클래스로 분류되며, 각 클래스는 숫자의 첫 자리에 의해 정의됩니다. 

**Informational responses (100 – 199)**

임시 응답으로 현재 클라이언트의 요청까지는 처리되었으니 계속 진행하라는 의미입니다. HTTP 1.1 버전부터 추가되었습니다. 

- 100 Continue
  - 클라이언트가 요청을 계속해야 한다는 것을 나타냅니다. 
  - 클라이언트는 요청 헤더에 `Expect: 100-continue`를 보내고 서버는 이를 처리할 수 있으면 이 코드로 응답합니다.
- 101 Switching Protocols
  - 서버가 클라이언트의 프로토콜 전환 요청을 수락했음을 나타냅니다. 
  - 프로토콜을 HTTP 1.1에서 업그레이드할 때 Upgrade 응답 헤더에 표시합니다. 
- 102 Processing
  - 서버가 처리하는 데 오랜 시간이 예상되어 클라이언트에서 타임 아웃이 발생하지 않도록 이 응답 코드를 보냅니다.

**Successful responses (200 – 299)**

요청이 성공적으로 처리되었음을 나타냅니다.

- 200 OK
  - 요청이 성공적으로 처리되었음을 나타냅니다.
- 201 Created
  - 요청이 성공적으로 처리되었으며, 새로운 리소스가 생성되었음을 나타냅니다. 
  - 응답 헤더 Location에 새로운 리소스의 URI를 기록합니다.
- 202 Accepted
  - 요청이 수락되었으나, 아직 처리되지 않았음을 나타냅니다.
  - 응답 헤더의 Location, Retry-After를 참고하여 클라이언트는 다시 요청을 보냅니다.
- 204 No Content
  - 요청이 성공적으로 처리되었으나, 반환할 내용이 없음을 나타냅니다.

**Redirection messages (300 – 399)**

클라이언트는 요청한 리소스를 다른 위치에서 가져와야 함을 나타냅니다.

- 300 Multiple Choices
  - 요청한 리소스에 대해 여러 가지 옵션이 있으며, 클라이언트가 선택할 수 있음을 나타냅니다. 
  - 예를 들어, 다른 포맷으로 리소스가 제공될 때 사용됩니다.
- 301 Moved Permanently
  - 요청한 리소스가 영구적으로 다른 위치로 이동했음을 나타냅니다.
  - 이동할 곳의 새로운 URI는 응답 헤더 Location에 기록합니다. 
  - 브라우저가 자동으로 캐싱하여 다음 요청부터는 서버를 거치지 않고 브라우저가 리다이렉션하여 변경된 url로 요청합니다. 
- 302 Found
  - 요청한 리소스가 임시로 다른 위치에 있음을 나타냅니다.
  - 302도 브라우저에서 캐싱이 가능하지만 대신 Cache-control이나 Expire 헤더를 명시해주어야 합니다. 
- 304 Not Modified
  - 클라이언트의 캐시된 리소스가 여전히 유효함을 나타냅니다.
  - If-Modified-Since와 같은 조건부 GET 요청일 때 지정한 리소스가 갱신되지 않았음을 알려줍니다. 

**Client error responses (400 – 499)**

클라이언트의 요청에 오류가 있음을 나타냅니다.

- 400 Bad Request
  - 잘못된 요청 구문으로 인해 서버가 요청을 이해할 수 없음을 나타냅니다.
- 401 Unauthorized
  - 요청한 리소스에 대한 액세스 권한이 없고, 인증이 필요함을 나타냅니다.
- 403 Forbidden
  - 서버가 요청을 거부했음을 나타냅니다. 권한이 없는 등의 상황에 사용합니다. 
  - 401과 다른 점은 서버가 클라이언트를 식별할 수 있다는 점입니다. 
- 404 Not Found
  - 요청한 리소스를 찾을 수 없음을 나타냅니다. 
- 408 Request Timeout
  - 서버가 요청을 시간 내에 받지 못했음을 나타냅니다.
- 429 Too Many Requests
  - 클라이언트가 주어진 시간 동안 너무 많은 요청을 보냈다.
  - 요청의 속도를 제한할 때 사용합니다. 응답에 Retry-After 헤더를 보내 얼마나 기다릴지를 알려 줄 수 있습니다.

**Server error responses (500 – 599)**

서버가 요청을 처리하는 도중에 오류가 발생했음을 나타냅니다.

- 500 Internal Server Error
  - 서버 내부 오류로 인해 요청을 처리할 수 없음을 나타냅니다.
- 501 Not Implemented
  - 서버가 요청된 기능을 지원하지 않음을 나타냅니다.
- 502 Bad Gateway
  - 게이트웨이 또는 프록시 역할을 하는 서버가 그 뒷단의 서버로부터 잘못된 응답을 받았음을 나타냅니다.
- 503 Service Unavailable
  - 서버가 일시적으로 과부하 상태이거나 유지보수 중임을 나타냅니다. 
- 504 Gateway Timeout
  - 게이트웨이 또는 프록시 서버가 그 뒷단의 서버로부터 지정된 시간 내에 응답을 받지 못했음을 나타냅니다.
